pkg/apis/pipeline/v1/task_validation.go

```diff
@@ -600,3 +600,39 @@ func validateTaskArraysIsolated(value, prefix string, arrayNames sets.String) *a
 func isParamRefs(s string) bool {
 	return strings.HasPrefix(s, "$("+ParamsPrefix)
 }
+
+// ValidateParamArrayIndex validates if the param reference to an array param is out of bound.
+// error is returned when the array indexing reference is out of bound of the array param
+// e.g. if a param reference of $(params.array-param[2]) and the array param is of length 2.
+// - `trParams` are params from taskrun.
+// - `taskSpec` contains params declarations.
+func (ts *TaskSpec) ValidateParamArrayIndex(ctx context.Context, params Params) error {
+	cfg := config.FromContextOrDefaults(ctx)
+	if cfg.FeatureFlags.EnableAPIFields != config.AlphaAPIFields {
+		return nil
+	}
+
+	// Collect all array params lengths
+	arrayParamsLengths := ts.Params.extractParamArrayLengths()
+	for k, v := range params.extractParamArrayLengths() {
+		arrayParamsLengths[k] = v
+	}
```

Comment:
NIT: much cleaner with maps.Copy(dest, source) 😄 

https://pkg.go.dev/golang.org/x/exp/maps#Copy

Reply:
Yes I didn't use it because we don't import that in the code base and I try to avoid changes like that